File: Medium/86 - PartitionList/Solution.cs
// Question 86. Partition List (https://leetcode-cn.com/problems/partition-list/)

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */

public class ListNode 
{
    public int val;
    public ListNode next;
    public ListNode(int x) { val = x; }
}

/// <summary>
/// 执行用时：100 ms, 在所有 C# 提交中击败了 100.00% 的用户
/// 内存消耗：25.1 MB, 在所有 C# 提交中击败了 6.06% 的用户
/// </summary>
public class Solution 
{
    public ListNode Partition(ListNode head, int x) 
    {
        ListNode newHead = new ListNode(-1);
        newHead.next = head;

        ListNode slow = newHead;
        ListNode fast = head;

        ListNode greaterListHead = new ListNode(-1);
        ListNode greaterList = greaterListHead;

        while (fast != null)
        {
            // System.Console.WriteLine($"slow = {slow.val}, fast = {fast.val}, greaterList = {greaterList.val}");

            ListNode next = fast.next;
            
            if (fast.val < x)
            {
                slow.next = fast;
                slow = slow.next;
                slow.next = null;
            }
            else
            {
                greaterList.next = fast;
                greaterList = greaterList.next;
                greaterList.next = null;
            }
            fast = next;
        }

        // System.Console.WriteLine($"slow = {slow.val}, fast = {fast}, greaterList = {greaterList.val}, greaterListHead.next = {greaterListHead.next.val}");
        slow.next = greaterListHead.next;

        return newHead.next;
    }
}

File: Medium/86 - PartitionList/solution.ts
// Question 86. Partition List (https://leetcode-cn.com/problems/partition-list/)

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

/**
 * 执行用时：84 ms, 在所有 TypeScript 提交中击败了 100.00% 的用户
 * 内存消耗：40.4 MB, 在所有 TypeScript 提交中击败了 57.14% 的用户
 *
 * Acknowledgement: This solution borrows idea from the [official solution](https://leetcode-cn.com/problems/partition-list/solution/fen-ge-lian-biao-by-leetcode-solution-7ade/)
 * 
 * @param {(ListNode | null)} head
 * @param {number} x
 * @returns {(ListNode | null)}
 */
function partition(head: ListNode | null, x: number): ListNode | null {
  const smallHead: ListNode = new ListNode(-1, head);
  let small: ListNode = smallHead;

  const largeHead: ListNode = new ListNode(-1, head);
  let large: ListNode = largeHead;

  while (head) {
    const next: ListNode | null = head.next;
    if (head.val < x) {
      small.next = head;
      small = small.next;
    } else {
      large.next = head;
      large = large.next;
    }
    head = next;
  }

  large.next = null;
  small.next = largeHead.next;
  return smallHead.next;
}
